workflow/controller/dag.go

```diff
@@ -87,7 +87,29 @@ func (d *dagContext) assessDAGPhase(targetTasks []string, nodes map[string]wfv1.
 			retriesExhausted = false
 		}
 	}
+
 	if unsuccessfulPhase != "" {
+		// If failFast set to false, we should return Running to continue this workflow for other DAG branch
+		if d.wf.Spec.FailFast != nil && !*d.wf.Spec.FailFast {
+			tmpOverAllFinished := true
+			// If all the nodes have finished, we should mark the failed node to finish overall workflow
+			// So we should check all the targetTasks have finished
+			for _, tmpDepName := range targetTasks {
+				tmpDepNode := d.getTaskNode(tmpDepName)
+				if tmpDepNode == nil {
+					tmpOverAllFinished = false
+					break
+				}
+				if tmpDepNode.Type == wfv1.NodeTypeRetry && hasMoreRetries(tmpDepNode, d.wf) {
+					tmpOverAllFinished = false
+					break
+				}
+			}
+			if !tmpOverAllFinished {
+				return wfv1.NodeRunning
+			}
+		}
```

Comment:
Do you mind writing a unit test for this?

Reply:
sure